pkg/localstorev2/internal/reserve/reserve.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package reserve
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/localstorev2/internal"
+	"github.com/ethersphere/bee/pkg/log"
+	storage "github.com/ethersphere/bee/pkg/storage"
+	storagev2 "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+)
+
+// loggerName is the tree path name of the logger for this package.
+const loggerName = "reserve"
+
+const storageRadiusKey = "reserve_storage_radius"
+const DefaultRadiusWakeUpTime = time.Minute * 5
+
+type Reserve struct {
+	mtx sync.Mutex
+
+	baseAddr     swarm.Address
+	stateStore   storage.StateStorer
+	radiusSetter topology.SetStorageRadiuser
+	logger       log.Logger
+
+	capacity int
+	size     int
+	radius   uint8
+}
+
+type Sample struct {
+	Items []swarm.Address
+	Hash  swarm.Address
+}
+
+/*
+	pull by 	bin - binID
+	evict by 	bin - batchID
+	sample by 	bin
+*/
+
+func New(baseAddr swarm.Address, store storagev2.Store, capacity int, reserveRadius uint8, stateStore storage.StateStorer, radiusSetter topology.SetStorageRadiuser, logger log.Logger) (*Reserve, error) {
+
+	rs := &Reserve{
+		baseAddr:     baseAddr,
+		capacity:     capacity,
+		stateStore:   stateStore,
+		radiusSetter: radiusSetter,
+		logger:       logger.WithName(loggerName).Register(),
+	}
+
+	var radius uint8
+	err := stateStore.Get(storageRadiusKey, &radius)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) { // fresh node
+			err = rs.SetRadius(reserveRadius)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	}
+
+	var size int
+	err = store.Iterate(storagev2.Query{
```

Comment:
So ideally if everything works as expected we should only have batchRadiusItems for the chunks we care about. So we should be able to just use `Count` for this and get the size right?

The reason I am saying this is tomorrow if we write a more optimum Count mechanism or use some other DB which provides one, we would do this much faster.